File: Data Structures & Algorithms/mergeSort/submission-0.cpp
// Definition for a Pair
// class Pair {
// public:
//     int key;
//     string value;
//
//     Pair(int key, string value) : key(key), value(value) {}
// };
class Solution {
public:
    vector<Pair> mergeSort(vector<Pair>& pairs) {
        if (pairs.size() <= 1) return pairs;

        int middle = pairs.size() / 2;
        vector<Pair> left(pairs.begin(), pairs.begin() + middle);
        vector<Pair> right(pairs.begin() + middle, pairs.end());
        left = mergeSort(left);
        right = mergeSort(right);
        int l = 0;
        int r = 0;
        vector<Pair> res;

        while (l < left.size() || r < right.size())
        {
            if (l == left.size())
            {
                res.push_back(right[r]);
                ++r;
            }
            else if (r == right.size())
            {
                res.push_back(left[l]);
                ++l;
            }
            else if (left[l].key <= right[r].key)
            {
                res.push_back(left[l]);
                ++l;
            }
            else
            {
                res.push_back(right[r]);
                ++r;
            }
        }

        return res;
    }
};

File: Data Structures & Algorithms/quickSort/submission-0.py
# Definition for a pair.
# class Pair:
#     def __init__(self, key: int, value: str):
#         self.key = key
#         self.value = value
class Solution:
    def quickSort(self, pairs: List[Pair]) -> List[Pair]:
        def qsort(pairs, s, e):
            if e - s + 1 <= 1:
                return pairs

            pivot = pairs[e]
            left = s
            
            for i in range(left, e):
                if pairs[i].key < pivot.key:
                    pairs[left], pairs[i] = pairs[i], pairs[left]
                    left += 1
            
            pairs[e] = pairs[left]
            pairs[left] = pivot

            pairs = qsort(pairs, s, left - 1)
            pairs = qsort(pairs, left + 1, e)

            return pairs
        
        return qsort(pairs, 0, len(pairs) - 1)
